File: sol/solution/0101-0200/0101/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).

-spec is_symmetric(Root :: #tree_node{} | null) -> boolean().
is_symmetric(Root) ->
    is_mirror(Root, Root).
is_mirror(null, null) -> true;
is_mirror(null, _) -> false;
is_mirror(_, null) -> false;
is_mirror(#tree_node{val = Val1, left = L1, right = R1}, #tree_node{val = Val2, left = L2, right = R2}) ->
    Val1 =:= Val2 andalso is_mirror(L1, R2) andalso is_mirror(R1, L2);
is_mirror(_, _) -> false.

File: sol/solution/0101-0200/0101/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
    is_mirror = ->(t1, t2) {
        return true if !t1 && !t2
        return false if !t1 || !t2
        t1.val == t2.val && is_mirror.call(t1.left, t2.right) && is_mirror.call(t1.right, t2.left)
    }
    is_mirror.call(root, root)
end
